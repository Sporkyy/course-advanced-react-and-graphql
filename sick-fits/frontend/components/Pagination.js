import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Head from 'next/head';
import Link from 'next/link';
import { perPage } from '../config';
import PaginationStyles from './styles/PaginationStyles';
import Error from './ErrorMessage';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ page }) => {
  return (
    <Query query={PAGINATION_QUERY}>
      {({ data, loading, error }) => {
        if (loading) return <p>Loading&hellip;</p>;
        if (error) return <Error error={error} />;
        const count = data.itemsConnection.aggregate.count;
        const pages = Math.ceil(count / perPage);
        return (
          <PaginationStyles data-test="pagination">
            <Head>
              <title>
                Sick Fits! ã€° Page {page} of {pages}
              </title>
            </Head>
            <Link
              prefetch
              href={{
                pathname: 'items',
                query: { page: page - 1 }
              }}
            >
              <a className="prev" aria-disabled={page <= 1}>
                &lt; Prev
              </a>
            </Link>
            <p>
              Page {page} of
              <span className="totalPages">{pages}</span>!
            </p>
            <p>{count} Items total</p>
            <Link
              prefetch
              href={{
                pathname: 'items',
                query: { page: page + 1 }
              }}
            >
              <a className="next" aria-disabled={pages <= page}>
                Next &gt;
              </a>
            </Link>
          </PaginationStyles>
        );
      }}
    </Query>
  );
};

export default Pagination;
export { PAGINATION_QUERY };
