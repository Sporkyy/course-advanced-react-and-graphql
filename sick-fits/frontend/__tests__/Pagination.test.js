import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';
import { MockedProvider } from 'react-apollo/test-utils';

Router.router = {
  push() {},
  prefetch() {}
};

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              __typename: 'count',
              count: length
            }
          }
        }
      }
    }
  ];
}

describe('<Pagination />', () => {
  it('Displays a loading message', () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    const pagination = wrapper.find('[data-test="pagination"]');
    expect(wrapper.text()).toContain('Loading');
  });

  it('Renders pagination for 18 items', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    // console.log(wrapper.find('.totalPages').debug());
    expect(wrapper.find('.totalPages').text()).toEqual('5');
    const pagination = wrapper.find('[data-test="pagination"]');
    expect(toJSON(pagination)).toMatchSnapshot();
    console.log(pagination.debug());
  });

  it('Disables prev button on first page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(true);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
  it('Disables next button on last page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(true);
  });
  it('Enables all buttonb on a middle page', async () => {
    const wrapper = mount(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.find('a.prev').prop('aria-disabled')).toEqual(false);
    expect(wrapper.find('a.next').prop('aria-disabled')).toEqual(false);
  });
});
