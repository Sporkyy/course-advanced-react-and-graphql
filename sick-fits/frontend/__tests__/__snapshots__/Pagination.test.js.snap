// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> Renders pagination for 18 items 1`] = `
Array [
  <PaginationStyles
    data-test="pagination"
  >
    <div
      className="PaginationStyles-aduuar-0 ewJsCc"
      data-test="pagination"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 0,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={true}
          className="prev"
          href="items?page=0"
          onClick={[Function]}
        >
          &lt; Prev
        </a>
      </Link>
      <p>
        Page 
        1
         of
        <span
          className="totalPages"
        >
          5
        </span>
        !
      </p>
      <p>
        18
         Items total
      </p>
      <Link
        href={
          Object {
            "pathname": "items",
            "query": Object {
              "page": 2,
            },
          }
        }
        prefetch={true}
      >
        <a
          aria-disabled={false}
          className="next"
          href="items?page=2"
          onClick={[Function]}
        >
          Next &gt;
        </a>
      </Link>
    </div>
  </PaginationStyles>,
  <div
    className="PaginationStyles-aduuar-0 ewJsCc"
    data-test="pagination"
  >
    <SideEffect(Head)>
      <Head />
    </SideEffect(Head)>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 0,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={true}
        className="prev"
        href="items?page=0"
        onClick={[Function]}
      >
        &lt; Prev
      </a>
    </Link>
    <p>
      Page 
      1
       of
      <span
        className="totalPages"
      >
        5
      </span>
      !
    </p>
    <p>
      18
       Items total
    </p>
    <Link
      href={
        Object {
          "pathname": "items",
          "query": Object {
            "page": 2,
          },
        }
      }
      prefetch={true}
    >
      <a
        aria-disabled={false}
        className="next"
        href="items?page=2"
        onClick={[Function]}
      >
        Next &gt;
      </a>
    </Link>
  </div>,
]
`;
